yt_dlp/extractor/bilibili.py

```diff
@@ -590,14 +740,54 @@ def _real_extract(self, url):
                 duration=traverse_obj(initial_state, ('videoData', 'duration', {int_or_none})),
                 __post_extractor=self.extract_comments(aid))
         else:
-            return {
-                **metainfo,
-                'duration': float_or_none(play_info.get('timelength'), scale=1000),
-                'chapters': self._get_chapters(aid, cid),
-                'subtitles': self.extract_subtitles(video_id, cid),
-                'formats': self.extract_formats(play_info),
-                '__post_extractor': self.extract_comments(aid),
-            }
+            formats = self.extract_formats(play_info)
+
+            if not traverse_obj(play_info, ('dash')):  # for legacy-only formats
+                has_qn = lambda x: x in traverse_obj(formats, (..., 'quality'))
+                for qn in traverse_obj(play_info, ('accept_quality', lambda _, v: not has_qn(v), {int})):
+                    formats.extend(traverse_obj(
+                        self.extract_formats(self._download_playinfo(video_id, cid, headers=headers, qn=qn)),
+                        (lambda _, v: not has_qn(v.get('quality')))))
+                self.check_missing_formats(play_info, formats)
+                if traverse_obj(formats, lambda _, v: v['fragments']):
+                    if not self._configuration_arg('_prefer_multi_flv'):
+                        # `_prefer_multi_flv` is mainly for writing test case, user should hardly need this
+                        dropping = ', '.join(traverse_obj(formats, (
+                            lambda _, v: v['fragments'], {lambda x: f'{x["format"]} ({x["format_id"]})'})))
+                        formats = traverse_obj(formats, lambda _, v: not v.get('fragments'))
+                        if dropping:
+                            self.to_screen(f'Dropping incompatible flv format(s) {dropping} when mp4 exists')
```

Comment:
There are two legacy format. One is single-file mp4 and the other is flv fragment(s).
Since flv formats needs a multi-video workaround, I couldn't figure out a way to have both formats for selection, and dropping flv formats looks like a better choice to me.

Reply:
Yeah, makes sense. If we wanted both we would need to implement a custom FileDownloader and I think for these flv legacy formats, that is just not worth it.
My questions was, what happens if there are no legacy mp4 formats, but legacy flv ones. Right now, all the flv formats would be dropped even though no mp4 formats exist, right?

Reply:
https://github.com/yt-dlp/yt-dlp/pull/9117/commits/d73d008ad4de4d0a4bee1a22aa9e1cc363911493
I see. I added additional check on line 754 so that dropping only happens when both mp4 and flv exist